### DISCUSSION TITLE
How does dtPathQueue work with only one dtNavMeshQuery?

### DISCUSSION THREAD

**Discussion by Git0Shuai** [recastnavigation/recastnavigation#590] [Q&A] [answered]
Hello. I'm trying to figure out how `dtPathQueue` works and find this in the `dtPathQueue::update()` function.

https://github.com/recastnavigation/recastnavigation/blob/d0b2ed8ff5d371f375cf49dad7ca96e8b22142d0/DetourCrowd/Source/DetourPathQueue.cpp#L111-L126

if there are more than one `dtPathQuery` InProgress, how to determine which should be updated in line 120 as all `dtPathQuery` share the same `m_navquery`?

dosen't every `dtPathQuery` in the queue need a corresponding `dtNavmeshQuery` instead of sharing the same one?

**Comment by grahamboree** [comment, marked as answer]
This code could use some comments and a bit of cleanup, but the general idea here is:

* There's a queue of path queries that we'd like to execute
* m_navquery holds the pathfinding state for the current path query
* `dtNavMeshQuery::updateSlicedFindPath` executes a certain number of iterations of the pathfinding process before saving its state and returning.  The maximum number of iterations to run is passed as the first parameter.  The number of iterations that were actually run is set in the second output parameter.
* dtPathQueue::update takes a maximum number of iterations to run and tries to get through as many path queries as it can.  If the first query that's executed consumes all the iterations, we break out of the loop at line 129 and don't execute any other path queries.  If it doesn't use all the iteration budget, the loop re-uses `m_navquery` to execute a new path search.

It's only ever executing one path query at at a time so it can re-use the same `dtNavMeshQuery` instance for subsequent queries.  The pathfinding results are stored in `PathQuery::path` in the call to `finalizeSlicedFindPath` on line 125, so it's not tossing the results when re-using `m_navquery`
